[NEW] Room's menu (REST API calls)

philipbrito opened the pull request:
@RocketChat/android

Closes #421 
Closes #422 
Closes #423 
Closes #356 

It also serves as a start point to add other functionalities to the app (for example: leave a room).

All the features are implemented using the [REST API calls](https://rocket.chat/docs/developer-guides/rest-api/).

**UPDATE**: #424 will not close because it is necessary to make changes to the corresponding API, although the most important stuff to implement this functionality on the Android app already is done by this PR.

## Progress

- [x] Show member list of a room (and its total of members).
     - [x] For DM
     - [x] For Private Channels (aka groups)
     - [x] For Public Channels (aka channels)
- [x] Show favorite messages list of a room.
     - [x] For DM
     - [x] For Private Channels (aka groups)
     - [x] For Public Channels (aka channels)
- [x] Show pinned messages of a room.
     - [x] For DM
     - [x] For Private Channels (aka groups)
     - [x] For Public Channels (aka channels)
- [ ] Show files list of a room. **Need REST API fixes**
     - [ ] For DM
     - [ ] For Private Channels (aka groups)
     - [ ] For Public Channels (aka channels)
- [x] Implement a lazy loading list.
- [x] Unit tests.

<img src="https://user-images.githubusercontent.com/11776825/30281513-05f9ff06-96e9-11e7-82c5-a475c870c819.jpg" alt="photo_1"> <img src="https://user-images.githubusercontent.com/11776825/30281517-0b73465e-96e9-11e7-9813-823ed7449125.jpg" alt="photo_1">

philipbrito committed:
Update app/build.gradle

philipbrito committed:
Update app/src/main/res/values/strings.xml

philipbrito committed:
Add app/src/main/res/menu/menu_room.xml

philipbrito committed:
Update activity_main.xml

philipbrito committed:
Update RoomFragment.java

philipbrito committed:
Delete app/src/main/res/layout-w720dp/fragment_room.xml

philipbrito committed:
Delete app/src/main/res/layout/fragment_room_main.xml

philipbrito committed:
Delete app/src/main/res/layout/room_side_menu.xml

philipbrito committed:
Update fa_strings.xml

philipbrito committed:
Delete UsersOfRoomDialogFragment.java

philipbrito committed:
Update RoomFragment.java

philipbrito committed:
Update fragment_room.xml

philipbrito commented on app/src/main/java/chat/rocket/android/fragment/chatroom/RoomFragment.java:
```diff
@@ -303,37 +299,40 @@ public boolean onItemLongClick(PairedMessage pairedMessage) {
-          .subscribe(
-              opened -> {
-                try {
-                  Field fieldSlidable = pane.getClass().getDeclaredField("mCanSlide");
```
Goodbye reflection!!!!!!!!!!

👋

philipbrito committed:
Update app/build.gradle

philipbrito committed:
Update app/build.gradle

philipbrito committed:
Add RoomDialogContract.kt

philipbrito committed:
Add RoomDialogFragment.kt

philipbrito committed:
Add RoomDialogPresenter.kt

philipbrito committed:
Update RocketChatAbsoluteUrl.java

philipbrito committed:
Update RoomFragment.java

philipbrito committed:
Update build.gradle

philipbrito committed:
Update RoomDialogContract.kt

philipbrito committed:
Update RoomDialogFragment.kt

philipbrito committed:
Update RoomDialogPresenter.kt

philipbrito committed:
Add RoomFileListAdapter.kt

philipbrito committed:
Add RoomMemberListAdapter.kt

philipbrito committed:
Add item_room_file.xml

philipbrito committed:
Add item_room_member.xml

philipbrito committed:
Add dialog_room.xml

philipbrito committed:
Add drawable

philipbrito committed:
Update RoomDialogPresenter.kt

philipbrito committed:
Update RoomDialogFragment.kt

philipbrito committed:
Update RoomFragment.java

philipbrito committed:
Update RoomDialogContract.kt

philipbrito committed:
RoomDialogFragment.kt

philipbrito committed:
RoomDialogPresenter.kt

philipbrito committed:
Update RoomDialogContract.kt

philipbrito committed:
Update RoomDialogFragment.kt

philipbrito committed:
Update RoomDialogPresenter.kt

philipbrito committed:
Add RoomPinnedMessagesAdapter.kt

philipbrito committed:
Uodate dialog_room.xml

philipbrito committed:
Update strings.xml

philipbrito committed:
Add RestApiHelper.kt

philipbrito committed:
Delete RoomDialogContract.kt

philipbrito committed:
Delete RoomDialogFragment.kt

philipbrito committed:
Delete RoomDialogPresenter.kt

philipbrito committed:
Add RoomListContract.kt

philipbrito committed:
Add RoomListFragment.kt

philipbrito committed:
Add RoomListPresenter.kt

philipbrito committed:
Delete dialog/RoomFileListAdapter.kt

philipbrito committed:
Delete dialog/RoomMemberListAdapter.kt

philipbrito committed:
Delete dialog/RoomPinnedMessagesAdapter.kt

philipbrito committed:
Delete dialog/RoomUserAdapter.java

philipbrito committed:
Delete dialog/RoomUserViewHolder.java

philipbrito committed:
Add list/RoomFileListAdapter.kt

philipbrito committed:
Add list/RoomMemberListAdapter.kt

philipbrito committed:
Add list/RoomPinnedMessagesAdapter.kt

philipbrito committed:
Add list/RoomUserAdapter.java

philipbrito committed:
Add list/RoomUserViewHolder.java

philipbrito committed:
Delete dialog_room.xml

philipbrito committed:
Add fragment_room_list.xml

philipbrito committed:
Update strings.xml

philipbrito committed:
Update MainActivity.java

philipbrito committed:
Update RoomFragment.java

philipbrito committed:
Update app/build.gradle

philipbrito committed:
Add RestApiHelperTest.kt

philipbrito committed:
Update AndroidManifest.xml

philipbrito committed:
Update MainActivity.java

philipbrito committed:
Add activity/room/RoomActivity.kt

philipbrito committed:
Update RestApiHelper.kt

philipbrito committed:
Update RoomListFragment.kt

philipbrito committed:
Update RoomListPresenter.kt

philipbrito committed:
Delete RoomPinnedMessagesAdapter.kt

philipbrito committed:
Add RoomMessagesAdapter.kt

philipbrito committed:
Delete list_item_message_newday.xml

philipbrito committed:
Add day.xml

philipbrito committed:
Update fragment_room_list.xml

philipbrito committed:
Update item_room_file.xml

philipbrito committed:
Update item_room_member.xml

philipbrito committed:
Update list_item_normal_message.xml

philipbrito committed:
Update list_item_system_message.xml

philipbrito committed:
Update colors.xml

philipbrito committed:
Update RoomFragment.java

philipbrito committed:
Update AbstractMessageViewHolder.java

philipbrito committed:
Add item_room_message.xml

philipbrito committed:
Add activity_room.xml

philipbrito committed:
Add message_style.xml

philipbrito committed:
Update RoomFileListAdapter.kt

philipbrito committed:
Update RoomMemberListAdapter.kt

philipbrito committed:
Update strings.xml

philipbrito committed:
Update EditMessageInteractorTest.kt

philipbrito committed:
Update PermissionInteractorTest.kt

philipbrito committed:
Delete message_styles.xml

philipbrito committed:
Merge branch 'develop' of https://github.com/RocketChat/Rocket.Chat.Android into iss424

philipbrito committed:
Update build.gradle

philipbrito committed:
Update RoomMemberListAdapter.kt

philipbrito committed:
Update RoomMessagesAdapter.kt

philipbrito committed:
Update OkHttpHelper.kt

philipbrito committed:
OkHttpHelper.kt

philipbrito committed:
Update RoomListFragment.kt

philipbrito committed:
Add UrlHelper.kt

philipbrito committed:
Update RoomListPresenter.kt

philipbrito committed:
Update RestApiHelper.kt

philipbrito committed:
Add UrlHelperTest.kt

philipbrito committed:
Update RoomListContract.kt

philipbrito committed:
Update RoomListFragment.kt

philipbrito committed:
Update RoomListPresenter.kt

philipbrito committed:
Update RoomFragment.java

philipbrito committed:
Update RoomMemberListAdapter.kt

philipbrito committed:
Update item_room_member.xml

philipbrito committed:
Update menu_room.xml

philipbrito committed:
Update message_style.xml

philipbrito committed:
Update DrawableHelper.kt

philipbrito committed:
Update User.java

rafaelks committed:
Merge branch 'develop' into iss424

rafaelks commented:
@filipedelimabrito Thanks! Looking good! 👍 

I did found a few UI issues that I believe we could improve (can be in another pull-request, no problem).

1. Members list cell row has too much space in the top; The space in the bottom & top should be the same IMHO;
2. The space between the member avatar & the status view (rounded colored) is too small, maybe something like 8pt would fit better; Here we could use the same horizontal space between status view & name/username label;
3. Favorited/Pinned messages doesn't show the image/attachment; I think we're not reusing the message cell here, I think we should;

Screenshots:

![screenshot_20170921-090111](https://user-images.githubusercontent.com/551004/30694896-66c79904-9eac-11e7-8968-136374dd7db0.png)
![screenshot_20170921-090736](https://user-images.githubusercontent.com/551004/30694906-6e24feb2-9eac-11e7-8132-8096843d4532.png)
![screen shot 2017-09-21 at 9 07 48 am](https://user-images.githubusercontent.com/551004/30694911-704bcc02-9eac-11e7-9899-17bb3074d5a3.png)

philipbrito committed:
Update RoomListContract.kt

philipbrito committed:
Update RestApiHelper.kt

philipbrito committed:
Update RoomListPresenter.kt

philipbrito committed:
Delete LoadMoreScrollListener.java

philipbrito committed:
Add LoadMoreScrollListener.kt

philipbrito committed:
Add EndlessRecyclerViewScrollListener.kt

More powerful. This class works if three types of RecyclerView.LayoutManager: LinearLayoutManager, GridLayoutManager and StaggeredGridLayoutManager.

rafaelks committed:
Merge branch 'develop' into iss424

philipbrito committed:
Update RoomListContract.kt

philipbrito committed:
Update RoomListFragment.kt

philipbrito committed:
Update RoomListPresenter.kt

philipbrito committed:
Update EndlessRecyclerViewScrollListener.kt

philipbrito committed:
Update RoomFileListAdapter.kt

philipbrito committed:
Update RoomMemberListAdapter.kt

philipbrito committed:
Update RoomMessagesAdapter.kt

philipbrito committed:
Update strings.xml

philipbrito committed:
Update item_room_member.xml

philipbrito committed:
Update item_room_message.xml

philipbrito committed:
Update item_room_member.xml

philipbrito committed:
Update item_room_message.xml

philipbrito committed:
Update RoomMessagesAdapter.kt

philipbrito committed:
Update RoomListFragment.kt

philipbrito commented:
(1) and (2): Fixed. I just changed the status view to another place on the screen. It is important to note that it follows the Material guidelines (as shown in the image below).

![photo_2017-09-22_23-40-00](https://user-images.githubusercontent.com/11776825/30769497-f50fc3fa-9fef-11e7-839d-049243e39286.jpg)

![photo_2017-09-22_23-40-03](https://user-images.githubusercontent.com/11776825/30769499-f909246a-9fef-11e7-9f37-9242093bde8d.jpg)

philipbrito commented:
(3): I added the widgets to show the images and attachments, but our REST API need a better design because there is a lot of loop to do to get a simple information, so it is bad.

We could improve that.

Example:

```
[...]
"urls":[
            {
               "url":"http:\/\/127.0.0.1:3000\/",
               "meta":{
                  "pageTitle":"Sign in | Patched Edition"
               },
               "headers":{
                  "contentType":"text\/html; charset=utf-8"
               },
               "parsedUrl":{
                  "host":"127.0.0.1:3000",
                  "hash":null,
                  "pathname":"\/users\/sign_in",
                  "protocol":"http:",
                  "port":"3000",
                  "query":null,
                  "search":null
               }
            },
[...]
```

philipbrito commented:
Also there is some missing information from the Rest: The user status, its name and offSet proprieties doesn't come from `/api/v1/channels.messages`, `/api/v1/groups.messages`  and `/api/v1/dm.messages`.

rafaelks commented:
Looking great @filipedelimabrito!!! Thanks!

I can reproduce one crash only: if I go to the list of members and tap "Back" before the list loads, it crashes after a few seconds to me.

philipbrito committed:
Update RoomListContract.kt

philipbrito committed:
Update RoomListFragment.kt

philipbrito committed:
Update RoomListPresenter.kt

philipbrito committed:
Update RoomMemberListAdapter.kt

philipbrito committed:
Update RoomMessagesAdapter.kt

philipbrito commented:
Thanks for reporting this crash @rafaels!

I just add some commits that fixes that! Also, I did some improvements in order to a better performance when viewing a large list (i.e. with many itens to show).

rafaelks committed:
Merge branch 'develop' into iss424

philipbrito commented:
#424 closed from #491
